Fix PasswordMixin.encrypted_columns

imtayadeway opened the pull request:
This mixin uses a class variable which has the (perhaps surprising)
behavior of sharing data up and down the inheritance
tree. Consequently, this leads to quirky behavior such as:

```
irb(main):001:0> Authentication.encrypted_columns
=> ["auth_key", "auth_key_encrypted", "password", "password_encrypted", "service_account", "service_account_encrypted"]
irb(main):002:0> MiqDatabase.encrypted_columns
=> ["auth_key", "auth_key_encrypted", "password", "password_encrypted", "service_account", "service_account_encrypted", "csrf_secret_token", "csrf_secret_token_encrypted", "session_secret_token", "session_secret_token_encrypted"]
irb(main):003:0> Authentication.encrypted_columns
=> ["auth_key", "auth_key_encrypted", "password", "password_encrypted", "service_account", "service_account_encrypted", "csrf_secret_token", "csrf_secret_token_encrypted", "session_secret_token", "session_secret_token_encrypted"]
```

You will notice here that the first invocation of
`Authentication.encrypted_columns` appears correct in returning the
encrypted columns as declared in that
model. `MiqDatabase.encrypted_columns` returns all its columns *plus*
those of `Authentication`. The second invokation of
`Authentication.encrypted_columns` now has `MiqDatabase`'s
concatenated to it.

~~The solution is to make this data an instance variable so that each
host has its own data.~~ The solution is to make this a `class_attribute` that's defined via
`ActiveSupport::Concern`'s `included`. This way, a base class such as
`Authentication` that includes this module can share that data with
its descendants, but that data isn't shared with other hosts that are
not kinds of `Authentication`

I've added tests for the two models mentioned above, however there are many  more subtypes of `Authentication` that are untested. This could possibly be done with the use of a shared example, but I'll leave that for a follow up PR as the tests I've provided are sufficient to expose the unexpected behavior.

@miq-bot add-label bug, core
@miq-bot assign @abellotti 

/cc @jrafanie

abellotti commented:
there was a reason for this, maybe subclasses I don't recall. API wasn't hiding all encrypted attributes without this, we'd need to run comparison tests for all collection classes, including reflections like config/settings/etc.

imtayadeway commented:
Here's a link to the original discussion: https://github.com/ManageIQ/manageiq/pull/2141#issuecomment-79023572. It looks like this introduces a regression....I'll try to find a way to resolve to the two issues. Marking as wip for now

@miq-bot add-label wip

Fryguy commented:
I think what you want is a [class_attribute](http://apidock.com/rails/Class/class_attribute)

imtayadeway commented:
@Fryguy thanks for the tip! Updated my PR and description to reflect the new change. I also added a new test to ensure that behavior for subclasses is preserved, using the example cited in https://github.com/ManageIQ/manageiq/pull/2141#issuecomment-79023572

@miq-bot rm-label wip
